chain/chain/src/chain.rs

```diff
@@ -754,7 +754,9 @@ impl Chain {
         // the block might not have been what the block producer originally produced. Either way, it's
         // OK if we miss some cases here because this is just an optimization to avoid reprocessing
         // known invalid blocks so the network recovers faster in case of any issues.
-        if error.is_bad_data() && !matches!(error, Error::InvalidSignature) {
+        if error.is_bad_data()
+            && !matches!(error, Error::InvalidSignature | Error::InvalidBlockHeight(_))
```

Comment:
Minor point but just checking, in our sequence of checks that we do for block, how far down the list do we have `InvalidBlockHeight`? Because if it's quite up in the stack of validation checks, this can be a source of spam as we don't mark such blocks as invalid.

Reply:
It's pretty early. However, the is_bad_data() check here is not for the purpose of preventing spam. It's for the purpose of dealing with a class of bugs where we the network gets stuck and the recovery is slowed down by repeatedly processing invalid blocks. Invalid height is not quite relevant here because these invalid blocks are talking about recent blocks. But honestly, I can't say that I still have enough context to reason about this. It's possible that this isn't even relevant anymore due to stateless validation. Either way, logically it is incorrect to mark a block as invalid just because its height is too high due to where the receiving node currently is, so we should stop doing that anyway.